### PR TITLE
State API

### PR DESCRIPTION
**Pull request by hackergrrl**
This introduces the `getState` function and the `state-update` event, as well as accompanying documentation and tests.

There's also some refactoring / cleanup that hopefully doesn't make the diff too hard to parse.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -107,67 +128,66 @@ Indexer.prototype._onNewFeed = function (feed, idx) {
...
   var didWork = false
 
-  var pending = 1
-
   // load state from storage
   if (!this._at) {
-    this._fetchState(function (err, state) {
+    this._fetchIndexState(function (err, state) {
+      // TODO: single error handling path (to emit + put in error state)
       if (err && !err.notFound) throw err // TODO: how to bubble up errors? eventemitter?
```
Comment (okdistribute) [inline index.js:191]: Do we want to emit the error here like we do in the other part of the file, using `onError`?

File: index.js [hunk 2]
```diff
@@ -202,6 +222,25 @@ Indexer.prototype._run = function () {
     var feeds = self._log.feeds()
     var nodes = []
 
+    // Check if there is anything new.
+    var indexedBlocks = Object.values(self._at).reduce((accum, entry) => accum + entry.max, 0)
```
Comment (okdistribute) [inline index.js:226]: Can we add a test for the indexer finishing successfully, and then adding new data so that it starts it up again, and adding pauses along the way (including when it's Idle?) Just wondering if pausing at any point in that process could cause concern.

File: index.js [hunk 3]
```diff
@@ -202,6 +222,25 @@ Indexer.prototype._run = function () {
...
+    var totalBlocks = self._log.feeds().reduce((accum, feed) => accum + feed.length, 0)
+
+    // Bail if no work needs to happen.
+    if (indexedBlocks === totalBlocks) {
+      return done()
```
Comment (okdistribute) [inline index.js:231]: would state be Idle here?

File: index.js [hunk 4]
```diff
@@ -238,7 +277,7 @@ Indexer.prototype._run = function () {
             didWork = true
             self._batch(nodes, function () {
               self._at[key].max += nodes.length
-              self._storeState(State.serialize(self._at, self._version), function () {
+              self._storeIndexState(IndexState.serialize(self._at, self._version), function () {
```
Comment (okdistribute) [inline index.js:280]: Would it be possible there is an error here that is uncaught?

File: index.js (hunk 1 above)
Comment (hackergrrl) [reply index.js:191]: Good catch!

File: index.js (hunk 2 above)
Comment (hackergrrl) [reply index.js:226]: In `test/pause-resume.js` it tests both pausing while idle, and pausing while indexing, and resuming.

File: index.js (hunk 3 above)
Comment (hackergrrl) [reply index.js:231]: Line 183 sets it to `PreIndexing` first, and then inside the `done` function it's set back to idle.

File: index.js (hunk 4 above)
Comment (hackergrrl) [reply index.js:280]: Yes! Same with `_batch` above. Good catch!
